Load balancer source ranges

justinsb opened the pull request:
This refactors #21431 to pull a lot of the code into cloudprovider so it
can be reused by AWS.

It also changes the name of the annotation to be non-GCE specific:
service.beta.kubernetes.io/load-balancer-source-ranges

Then also adds a null-implementation for OpenStack

justinsb committed:
AWS: Add support for load balancer source ranges

This refactors #21431 to pull a lot of the code into cloudprovider so it
can be reused by AWS.

It also changes the name of the annotation to be non-GCE specific:
service.beta.kubernetes.io/load-balancer-source-ranges

Fix #21651

justinsb committed:
Openstack null-support for load balancer source

We return an error if the user specifies a non 0.0.0.0/0 load balancer
source restriction on OpenStack, where we can't enforce the restriction
(currently).

freehan commented:
Thanks for putting this together.

Since the annotation becomes generic, maybe it is better to validate it upon service creation. So that users do not need to check service events to see the validation error. Instead, error will show up immediately upon creation. Plus avoiding failing retries. 

This can be done at  
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L102
or
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L1634

justinsb commented:
(fixed gofmt)

justinsb commented:
Question: I don't know if this applies to other clouds, but on AWS there's the concept of an internal ELB.  You might want to protect access to an internal ELB by the security-group that an instance is in (rather than the IP).  Should we make the label `service.beta.kubernetes.io/load-balancer-source-restrictions` or something that encompasses this use-case?  @jsravn is thinking about supporting internal ELBs; maybe if you specify internal source restrictions it becomes an internal ELB.

I think the equivalent for GCE might be restricting on the instance tags (?)

cc @jsravn

justinsb commented:
@freehan I'm not sure I'm going to have time in the 1.2 timeframe to add validation on creation.  I think it's important to get this in to 1.2 so we have the same annotation name across cloudproviders.  Otherwise it'll be a doc/support nightmare...

@thockin this is pretty important for 1.2 for AWS, partially because I don't think it's a good idea to ship with per-provider annotations that do the same thing, though primarily because I (stupidly) derived a bunch of bug-fix PRs from it.  Can you have a look and make a call on whether we should put this in to 1.2?  If it's a no, that's fine, but I then need to rebase a bunch of more critical PRs.

justinsb commented:
(rebased)

thockin commented on pkg/cloudprovider/cloud.go:
```diff
@@ -167,3 +167,7 @@ func (s ServiceAnnotation) GetValue(key string) (string, bool) {
 }
+
+const (
+	LBAnnotationAllowSourceRange = "service.beta.kubernetes.io/load-balancer-source-ranges"
```
comment on the expected value format

thockin replied:
This sort of leaves NodePorts in limbo - is that a problem?  Or conversely, once we have proper network policy, does this become obsolete?

justinsb replied:
Not sure why it leaves NodePorts in limbo.  Do you mean because we don't automatically configure them with a comparable annotation?  We don't currently configure any access for NodePorts.  So I think that is fine ... I see NodePorts as a building block e.g. for configuring a hardware load balancer, or for use for federated cloud interconnectivity.  I would imagine the component building on NodePorts would do whatever additional configuration it had to do, and so we should add extra support for NodePorts when we need it.

That said, I'm not sure exactly what motivated the addition of this annotation for GCE in the first place (I see the utility, just not sure whether there was something in particular because we didn't tie the original PR back to an issue)

thockin replied:
I meant "limbo" because the new net policy stuff applies policy to the underlying service, and this applies policy to the LB frontend, leaving NodePort as the only `type` that doesn't have some policy enforcement.  We *could* apply policy to node ports, it wouldn't even be that hard.  I don't think we need to, I just wanted to think about it.

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+	defaultLBSourceRange = "0.0.0.0/0"
+)
+
+// TODO: Should I move this into utils/sets ?  Even though it isn't generated?
```
I don't think so.  This has a custom-enough API that we would not want to templatize it, I think.

justinsb replied:
OK ... will remove comment.  We can always move later if we add sets support for non-map-key types

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+	return a
+}
+
+// Equal checks if to IPNetSets are equal (ignoring order)
```
s/to/two

justinsb replied:
Ugh... so embarrassing.

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+
+// IsAllowAll checks whether the IPNetSet contains the default allow-all policy
+func IsAllowAll(ipnets IPNetSet) bool {
+	if ipnets.Len() != 1 {
```
is it not valid (weird but valid) to provide "10.0.0.0/8,0.0.0.0/0" ?

justinsb replied:
Definitely weird, but definitely valid.  I don't think the user can be too upset if the OpenStack provider complains about that policy.  But IsAllowAll will probably get used in future in different ways...

thockin replied:
IsAllowAll() should return true here, no?

On Mon, Feb 29, 2016 at 4:34 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> In pkg/cloudprovider/utils.go
> <https://github.com/kubernetes/kubernetes/pull/21907#discussion_r54505274>
> :
>
> > +	val := annotation[LBAnnotationAllowSourceRange]
> > +	val = strings.TrimSpace(val)
> > +	if val == "" {
> > +		val = defaultLBSourceRange
> > +	}
> > +	specs := strings.Split(val, ",")
> > +	ipnets, err := ParseIPNetSet(specs)
> > +	if err != nil {
> > +		return nil, fmt.Errorf("Service annotation %s:%s is not valid. Expecting a comma-separated list of source IP ranges. For example, 10.0.0.0/24,192.168.2.0/24", LBAnnotationAllowSourceRange, val)
> > +	}
> > +	return ipnets, nil
> > +}
> > +
> > +// IsAllowAll checks whether the IPNetSet contains the default allow-all policy
> > +func IsAllowAll(ipnets IPNetSet) bool {
> > +	if ipnets.Len() != 1 {
>
> Definitely weird, but definitely valid. I don't think the user can be too
> upset if the OpenStack provider complains about that policy. But IsAllowAll
> will probably get used in future in different ways...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21907/files#r54505274>.
>

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+		return false
+	}
+	s := ipnets.StringSlice()
+	return s[0] == "0.0.0.0/0"
```
What if it is not [0]?  You don't sort the string slice..

justinsb replied:
length == 1 here.  But I'm going to have to fix that because of the "weird" edge case.  The real goal here was not to break OpenStack if the user specified "0.0.0.0/0" (the one policy OpenStack can support, and also the default)

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+	specs := strings.Split(val, ",")
+	ipnets, err := ParseIPNetSet(specs)
+	if err != nil {
+		return nil, fmt.Errorf("Service annotation %s:%s is not valid. Expecting a comma-separated list of source IP ranges. For example, 10.0.0.0/24,192.168.2.0/24", LBAnnotationAllowSourceRange, val)
```
If this is really a beta API, why not validate the annotation in the API validation logic, so invalid values could never make it through?

justinsb replied:
OK, seeing as you both requested it :-)

One problem is that I'm going to have to move some stuff to keep the package dependencies clean

thockin commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -790,7 +764,15 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
+	if err != nil {
+		// This really shouldn't happen... GCE has returned something unexpected
+		glog.Warningf("Error parsing firewall SourceRanges: %v", fw.SourceRanges)
+		// We don't return the error, because we can hopefully recover from this by reconfiguring the fireall
```
firewall typo

justinsb replied:
Oops

thockin commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,99 @@
+
+// GetSourceRangeAnnotations verifies and parses the LBAnnotationAllowSourceRange annotation from a service,
+// extracting the source ranges to allow, and if not present returns a default (allow-all) value.
+func GetSourceRangeAnnotations(annotation ServiceAnnotation) (IPNetSet, error) {
```
I still see no value in the `ServiceAnnotation` type - I think we should get rid of it and just call it `map[string]string`

justinsb replied:
Agreed

thockin commented:
holla when ready for round 2

justinsb committed:
LB Source Ranges: Move validation into API layer

Had to move other things around too to avoid a weird api ->
cloudprovider dependency.

Also adding fixes per code reviews.

(This is a squash of the previously approved commits)

justinsb commented:
@thockin ready for review round 2 I think.  Had to rearrange a lot of code to get validation in the API layer to be clean.

justinsb commented:
Couldn't resist adding another test...

justinsb commented:
Added a test for the service validation also!

thockin commented on pkg/util/sets/ipnet.go:
```diff
@@ -0,0 +1,119 @@
+	"strings"
+)
+
+type IPNet map[string]*net.IPNet
```
I am not sure how I feel about having an open-coded set implementation among all the auto-generated ones, especially since it has a different form and slightly expanded API.

@lavalamp thoughts?  Other options could include putting this into `util/net` or autogenerating it, and forcing callers to call List() and then convert to strings themselves, or even standardizing StringList() in generated sets (printf("%v") or allow some customization in set-gen's input).

lavalamp replied:
Hm... Yeah this is not ideal. Can the interface at least be made consistent, so it will be easy to replace with an autogenerated one in the future?

justinsb replied:
I think it is consistent, though not feature complete.  But I'm probably missing something?

Can we make a decision here?  This PR is blocking a bunch of other PRs and that set of PRs are probably my only blockers for cutting the branch as far as I'm concerned for 1.2 (which isn't the sole consideration, just sharing the view from my "station").  I'm happy to leave this here and harmonize the interface (e.g. adding missing methods).  Or I'm happy to move this anywhere else!

thockin replied:
I think we should move it to a different pkg with a big comment to fold it
back into autogenerated sets.

What do you mean that IPNet can't be used as a map key?

http://play.golang.org/p/xLDdbzvUUS

On Tue, Mar 1, 2016 at 5:07 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> In pkg/util/sets/ipnet.go
> <https://github.com/kubernetes/kubernetes/pull/21907#discussion_r54664583>
> :
>
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package sets
> > +
> > +import (
> > +	"net"
> > +	"strings"
> > +)
> > +
> > +type IPNet map[string]*net.IPNet
>
> I think it is consistent, though not feature complete. But I'm probably
> missing something?
>
> Can we make a decision here? This PR is blocking a bunch of other PRs and
> that set of PRs are probably my only blockers for cutting the branch as far
> as I'm concerned for 1.2 (which isn't the sole consideration, just sharing
> the view from my "station"). I'm happy to leave this here and harmonize the
> interface (e.g. adding missing methods). Or I'm happy to move this anywhere
> else!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21907/files#r54664583>.
>

justinsb replied:
OK I'll move it.  Shadowing net is naughty (util/net), but it looks like the best place.  But if we are shadowing anyway, how about putting it under net (util/net/sets), so it will be netsets.IPNet?

Pretty sure it can't be used as a map key, because you can't use anything with []byte as a map key, and IP is a []byte - the workaround is to use a string representation.

With pointers retrieval doesn't work: http://play.golang.org/p/Mc1BO9T5GH

Trying to use `map[net.IPNet]struct{}` gives `invalid map key type net.IPNet`: http://play.golang.org/p/qrw1UvQ6tH

Another option would be to create a string-backed IPNet e.g. http://play.golang.org/p/Mbp-AH-Isb  But really this should just be fixed at the language level...

thockin commented on pkg/api/annotations.go:
```diff
@@ -0,0 +1,44 @@
+
+// GetServiceLoadBalancerSourceRanges verifies and parses the ServiceAnnotationLoadBalancerSourceRangesKey annotation from a service,
+// extracting the source ranges to allow, and if not present returns a default (allow-all) value.
+func GetServiceLoadBalancerSourceRanges(annotations map[string]string) (sets.IPNet, error) {
```
what about hosting this in pkg/api/service/util.go (see ...endpoints/util.go and .../pod/util.go for precedent).  The name can get less mouthy, too.  `service.GetLoadBalancerSourceRanges()`

justinsb replied:
I like that idea...

justinsb commented on pkg/cloudprovider/utils.go:
```diff
@@ -0,0 +1,29 @@
+import "k8s.io/kubernetes/pkg/util/sets"
+
+// IsAllowAll checks whether the sets.IPNet allows traffic from 0.0.0.0/0
+func IsAllowAll(ipnets sets.IPNet) bool {
```
Moving it.

I'd rather keep it as a function if that's OK, because I'm imagining IPv6 support not being too far down the road...  AWS allows IPv6 on ELBs (but not natively on their internal network other than through tunnelling) - it's much easier to add than making the whole internal network IPv6 (much to my dismay).

justinsb commented:
Pushed fixes per code review.

lavalamp commented:
Gah! Tim, your example has a big bug which I'll demonstrate shortly

lavalamp commented:
OK, Justin beat me to it. :)

thockin commented on pkg/api/types.go:
```diff
@@ -1411,6 +1411,15 @@ const (
+	// access only from the CIDRs currently allocated to MIT & the USPS.
+	//
+	// Not all cloud providers support this annotation, though AWS & GCE do.
+	ServiceAnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
```
what would you think about putting this into pkg/api/service/* - It's not really part of the versioned API.. thoughts?

justinsb replied:
I think that makes sense - I'll do it.

justinsb replied:
We do have "unversioned/well_known_labels.go", which is perhaps even closer (except for the eternal labels vs annotations debate).  Where would you prefer?

thockin replied:
pkg/api/service

On Wed, Mar 2, 2016 at 3:13 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/21907#discussion_r54809534>
> :
>
> > @@ -1411,6 +1411,15 @@ const (
> >  	// ClusterIPNone - do not assign a cluster IP
> >  	// no proxying required and no environment variables should be created for pods
> >  	ClusterIPNone = "None"
> > +
> > +	// ServiceAnnotationLoadBalancerSourceRangesKey is the key of the annotation on a service to set allowed ingress ranges on their LoadBalancers
> > +	//
> > +	// It should be a comma-separated list of CIDRs, e.g. `0.0.0.0/0` <http://0.0.0.0/0> to
> > +	// allow full access (the default) or `18.0.0.0/8,56.0.0.0/8` <http://18.0.0.0/8,56.0.0.0/8> to allow
> > +	// access only from the CIDRs currently allocated to MIT & the USPS.
> > +	//
> > +	// Not all cloud providers support this annotation, though AWS & GCE do.
> > +	ServiceAnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
>
> We do have "unversioned/well_known_labels.go", which is perhaps even
> closer (except for the eternal labels vs annotations debate). Where would
> you prefer?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21907/files#r54809534>.
>

thockin commented on pkg/api/validation/validation_test.go:
```diff
@@ -2877,6 +2877,20 @@ func TestValidateService(t *testing.T) {
+		{
+			name: "invalid LoadBalancer source range annotation",
+			tweakSvc: func(s *api.Service) {
+				s.Annotations[api.ServiceAnnotationLoadBalancerSourceRangesKey] = "1.2.3.4/33"
```
testcase for = "" and = "foo.bar", please ?

thockin commented:
That's all I have left to nitpick.

justinsb commented:
Just waiting on travis, but I think I've addressed the comments @thockin 

And I even found some nits of my own ;-)

eparis commented:
ok to test
@k8s-bot test this issue: #IGNORE

justinsb commented:
Pushed a squashed commit... it needed a squash anyway, and hopefully this will wake Jenkins up.  Will re-add LGTM in a minute.

eparis commented:
busted again....

ixdy commented:
@k8s-bot test this, issue #22435

ixdy commented:
yay, actually triggered!

eparis commented:
failed, but triggered  :)

justinsb commented:
@k8s-bot unit test this github flake #19403

caesarxuchao commented:
@justinsb do you happen to keep the link to the test logs of the 19403 failure?

ixdy commented:
@caesarxuchao you should be able to find them under http://pr-test.k8s.io/21907

caesarxuchao commented:
That's handy. Thanks @ixdy!
